File: cloud-collect-core/src/main/java/com/windea/demo/cloudcollect/core/component/StringToSortConverter.java
package com.windea.demo.cloudcollect.core.component;

import org.springframework.core.convert.converter.Converter;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Component;
import org.springframework.util.StringUtils;

import java.util.Arrays;
import java.util.stream.Collectors;

/**
 * 字符串到排序对象的转化器。
 * <p>示例格式："", "+name", "+name,-age", "name,age"。
 */
@Component
public class StringToSortConverter implements Converter<String, Sort> {
	@Override
	public Sort convert(String string) {
		var orders = Arrays.stream(string.split(","))
			.map(String::trim).filter(StringUtils::hasText)
			.map(s -> {
				if(s.startsWith("+")) {
					return Sort.Order.asc(s.substring(1));
				} else if(s.startsWith("-")) {
					return Sort.Order.desc(s.substring(1));
				} else {
					return Sort.Order.asc(s);
				}
			})
			.collect(Collectors.toList());

		return Sort.by(orders);
	}
}

File: cloud-collect-demo/src/main/kotlin/com/windea/demo/cloudcollect/demo/component/StringToSortConverter.kt
package com.windea.demo.cloudcollect.demo.component

import org.springframework.core.convert.converter.*
import org.springframework.data.domain.*
import org.springframework.stereotype.*

/**
 * 字符串到排序对象的转换器。不检查要排序的属性的正确性。
 *
 * 示例格式："", "+name", "+name,-age", "name,age"。
 */
@Component
class StringToSortConverter : Converter<String, Sort> {
	override fun convert(string: String): Sort {
		val orders = string.split(",").filter { it.isNotBlank() }
			.map { s ->
				when {
					s.startsWith("+") -> Sort.Order.asc(s.substring(1))
					s.startsWith("-") -> Sort.Order.desc(s.substring(1))
					else -> Sort.Order.asc(s)
				}
			}
		return Sort.by(orders)
	}
}
